c++/src/kj/common.h

```diff
@@ -2098,6 +2098,49 @@ _::Deferred<Func> defer(Func&& func) {
 #define KJ_DEFER(code) auto KJ_UNIQUE_NAME(_kjDefer) = ::kj::defer([&](){code;})
 // Run the given code when the function exits, whether by return or exception.
 
+// =======================================================================================
+// IsDisallowedInCoroutine
+
+namespace _ {
+
+template <typename T, typename = void>
+struct IsDisallowedInCoroutine {
+  static constexpr bool value = false;
+};
+
+template <typename T>
+struct IsDisallowedInCoroutine<T, typename Decay<T>::_kj_DissalowedInCoroutine> {
+  static constexpr bool value = true;
+};
+
+template <typename T>
+struct IsDisallowedInCoroutine<T*, typename T::_kj_DissalowedInCoroutine> {
+  static constexpr bool value = true;
+};
+
+template <typename T>
+constexpr bool isDisallowedInCoroutine() {
+  return IsDisallowedInCoroutine<T>::value;
+}
+
+}  // namespace _
+
+#define KJ_DISALLOW_AS_COROUTINE_PARAM \
+  using _kj_DissalowedInCoroutine = void; \
+  template<typename T> friend constexpr bool kj::_::isDisallowedInCoroutine(); \
```

Comment:
Nitpick: I think these friend declarations aren't necessary unless we're expecting the macro to be used in the `private:` section, but I'd argue it should be under `public:` anyway.

Reply:
Yep, that's exactly why I added them. I'd prefer to keep them as I'd prefer folks to not have to think much about whether the macro should be public or private.